pipeval/validate/validators/fastq.py

```diff
@@ -0,0 +1,120 @@
+        quality = record[3]
+
+        if not FASTQ_RECORD_VALIDATOR.record_identifier_format.match(identifier):
+            invalid_entries.append(
+                f'Record identifier `{identifier}` is invalid. It must begin with \'@\''
+            )
+
+        if not FASTQ_RECORD_VALIDATOR.sequence_format.match(sequence):
+            invalid_entries.append(
+                f'Read sequence `{sequence}` contains invalid characters. '
+                'Only \'ACTGNactgn\' are allowed'
+            )
+
+        if not FASTQ_RECORD_VALIDATOR.extra_field_format.match(extra_field):
+            invalid_entries.append(
+                f'Extra field `{extra_field}` is invalid. It must begin with \'+\''
+            )
+
+        if len(sequence) != len(quality):
+            invalid_entries.append(
+                f'Sequence and quality must be of the same length:\n'
+                f'Sequence - {len(sequence)} - {sequence}\n'
+                f'Quality - {len(quality)} - {quality}'
+            )
+
+        min_quality = min(quality)
+        max_quality = max(quality)
+
+        if ord(min_quality) < FASTQ_RECORD_VALIDATOR.minimum_quality_ordinal or \
+            ord(max_quality) > FASTQ_RECORD_VALIDATOR.maximum_quality_ordinal:
+            invalid_entries.append(
+                f'Quality scores out of valid range. Quality scores must fall between '
+                f'{FASTQ_RECORD_VALIDATOR.minimum_quality_ordinal} and '
+                f'{FASTQ_RECORD_VALIDATOR.maximum_quality_ordinal}. '
+                f'Found min:`{min_quality}` ({ord(min_quality)}) and '
+                f'max:`{max_quality}` ({ord(max_quality)}) in record.'
+            )
+
+        if len(invalid_entries) != 0:
+            record_errors = '\n'.join(invalid_entries)
+            raise ValueError(f'Record {record} is invalid: {record_errors}')
+
+# pylint: disable=R0903
+class FASTQ():
+    ''' FASTQ file handling and validation class '''
+    def __init__(self, fastq_file:Path):
+        '''Constructor'''
+        self._fastq_path = fastq_file
+        self._file_handler = self._get_file_handler()
+
+    def _get_file_handler(self):
+        '''Detect file format and return approriate handler to read file'''
+        _handler_map = {
+            'application/x-gzip': gzip.open,
+            'application/x-bzip2': bz2.open,
+            'text/plain': open
+        }
+
+        file_mime = magic.from_file(self._fastq_path.resolve(), mime=True)
+
+        handler = _handler_map.get(file_mime, None)
+
+        if handler is None:
+            raise TypeError(f'Unexpected FASTQ format `{file_mime}` for file: `{self._fastq_path}`')
+
+        return handler
+
+    def validate_fastq(self):
+        ''' Validate the FASTQ file '''
+        with self._file_handler(self._fastq_path, 'rt') as rd:
+            current_record = []
+            current_record_length = 0
+            for line in rd:
+                current_record.append(line.strip())
```

Comment:
Looks good. In the future, consider taking advantage of methods like `enumerate` and `range`, which can make some looping operations more concise. Here are a couple examples of how it might be done here.

```python
current_record = []
with open("foo", "r") as rd:
    for (i,line) in enumerate(rd):
        if i >= 4:
            continue
        current_record.append(line.strip())

```
or
```python
current_record = []
with open("foo", "r") as rd:
    current_record = [x.strip() for x in [next(rd, None) for line in range(4)] if x]
```

Reply:
Got it, I'll keep those in mind for the future!